packages/firestore/src/index/firestore_index_value_writer.ts

```diff
@@ -0,0 +1,197 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DocumentKey } from '../model/document_key';
+import { normalizeByteString, normalizeNumber } from '../model/normalize';
+import { isMaxValue } from '../model/values';
+import { ArrayValue, MapValue, Value } from '../protos/firestore_proto_api';
+import { fail } from '../util/assert';
+import { isNegativeZero } from '../util/types';
+
+import { DirectionalIndexByteEncoder } from './directional_index_byte_encoder';
+
+// Note: This code is copied from the backend. Code that is not used by
+// Firestore was removed.
+
+const INDEX_TYPE_NULL = 5;
+const INDEX_TYPE_BOOLEAN = 10;
+const INDEX_TYPE_NAN = 13;
+const INDEX_TYPE_NUMBER = 15;
+const INDEX_TYPE_TIMESTAMP = 20;
+const INDEX_TYPE_STRING = 25;
+const INDEX_TYPE_BLOB = 30;
+const INDEX_TYPE_REFERENCE = 37;
+const INDEX_TYPE_GEOPOINT = 45;
+const INDEX_TYPE_ARRAY = 50;
+const INDEX_TYPE_MAP = 55;
+const INDEX_TYPE_REFERENCE_SEGMENT = 60;
+
+// A terminator that indicates that a truncatable value was not truncated.
+// This must be smaller than all other type labels.
+const NOT_TRUNCATED = 2;
+
+/** Firestore index value writer.  */
+export class FirestoreIndexValueWriter {
+  static INSTANCE = new FirestoreIndexValueWriter();
+
+  private constructor() {}
+
+  // The write methods below short-circuit writing terminators for values
+  // containing a (terminating) truncated value.
+  //
+  // As an example, consider the resulting encoding for:
+  //
+  // ["bar", [2, "foo"]] -> (STRING, "bar", TERM, ARRAY, NUMBER, 2, STRING, "foo", TERM, TERM, TERM)
+  // ["bar", [2, truncated("foo")]] -> (STRING, "bar", TERM, ARRAY, NUMBER, 2, STRING, "foo", TRUNC)
+  // ["bar", truncated(["foo"])] -> (STRING, "bar", TERM, ARRAY. STRING, "foo", TERM, TRUNC)
+
+  /** Writes an index value.  */
+  writeIndexValue(value: Value, encoder: DirectionalIndexByteEncoder): void {
+    this.writeIndexValueAux(value, encoder);
+    // Write separator to split index values
+    // (see go/firestore-storage-format#encodings).
+    encoder.writeInfinity();
+  }
+
+  private writeIndexValueAux(
+    indexValue: Value,
+    encoder: DirectionalIndexByteEncoder
+  ): void {
+    if ('nullValue' in indexValue) {
```

Comment:
optional: use `switch` here, since every if-condition is a single statement.

Reply:
Unfortunately, switch does not work here since we are not doing an equals comparison on a value.